### ISSUE TITLE
Proposal: resilience4j-micrometer integration should use tags

### ISSUE THREAD

**Issue by noroute** [resilience4j/resilience4j#354]
# Context
Micrometer regards dimensions as a first-class citizen of time series and advises to not include them in the time series name (see https://micrometer.io/docs/concepts#_meters).

resilience4j-micrometer adds the name of CircuitBreakers (or Bulkheads, Retries and RateLimits for that matter) to the metric time series. This leads to limitations for querying (e.g. "find me all the open circuit breakers" would not be possible) and a large number of series being published (e.g. 6x #circuitbreakers) which is not optimal for most time series databases.

# Proposal
With this issue I propose to use tags for dimensions for all the Metrics currently supported (CircuitBreaker, RateLimit, Bulkhead, Retry).

# Consequences
The implementation changes would be trivial (see [Gauge Fluent Builder](https://micrometer.io/docs/concepts#_gauge_fluent_builder)) but would result in *backward incompatible change* in structure and names of time series. I am not sure if this is ok or if we should add this as a non-default option. I would be able to contribute a PR if we get agreement on an option.

# Example
I use CircuitBreaker as an example here as it has the most complex structure of the four.

## Current state
In Prometheus, this is what you get by default (`cb1` being the circuitbreaker name, only showing one metric):
```
# HELP resilience4j_circuitbreaker_cb1_successful  
# TYPE resilience4j_circuitbreaker_cb1_successful gauge
resilience4j_circuitbreaker_cb1_successful 0.0
```

## Proposed state: tags for names and metrics
To make discovery and drilldown easier, `name` and `metric` should be added as a dimension. This would allow drilldown on one specific circuitbreaker:
```
# HELP resilience4j_circuitbreaker  
# TYPE resilience4j_circuitbreaker gauge
resilience4j_circuitbreaker_stats{name="cb1",metric="successful"} 0.0
resilience4j_circuitbreaker_state{name="cb1"} 3.0
```

This version would support queries like:
* list of all open CircuitBreakers
* overview of buffered calls across all CircuitBreakers
* CircuitBreakers with more than 300 failed calls

**Comment by RobWin** [comment]
Thx for this well structured issue :1st_place_medal: 
I totally agree.  But as you said it's a backward incompatible change and the resilience4j-prometheus module should get a new major version, if we do semantic versioning. 
We could start to give every module it's own version? Or should we start to separate add-on modules from the main project? But then the release process would be really cumbersome.

We could also think about creating a second major version of our resilience4j-prometheus module -> `resilience4j-prometheus2` ? 

What do you think @storozhukBM?

**Comment by noroute** [comment]
I just stumbled upon the docs for [Spring Boot](https://github.com/resilience4j/resilience4j/blob/master/resilience4j-documentation/src/docs/asciidoc/addon_guides/springboot.adoc)/[Spring Boot 2](https://github.com/resilience4j/resilience4j/blob/master/resilience4j-documentation/src/docs/asciidoc/addon_guides/springboot2.adoc) integration and they have examples that look *almost exactly* like what I proposed. I tried version 0.13.2 (Spring Boot 2 integration + Prometheus) and what I got was the mentioned example. This is also consistent with the source code. What's going on here?

**Comment by noroute** [comment]
Ok, after looking a bit more it seems like `resilience4j-prometheus` takes the "dimensions as tags" approach but `resilience4j-micrometer` does *not*. The examples with Spring Boot all use `resilience4j-prometheus` while Micrometer is the official Spring Boot metrics adapter starting with Boot 2.0. Which means that direct prometheus and micrometer integrations behave inconsistently.

What's the way forward? The proposed change would only be breaking for `resilience4j-micrometer`, not for the others (AFAICS).

**Comment by RobWin** [comment]
If we follow the metric and label conventions of Prometheus, we should even rename our metrics.

> ...should have a (single-word) application prefix relevant to the domain the metric belongs to. The prefix is sometimes referred to as namespace by client libraries.

In our case the namespace is:  `resilience4j_circuitbreaker` or `resilience4j_ratelimiter`.

> Every time series is uniquely identified by its metric name. The metric name specifies the general feature of a system that is measured (e.g. http_requests_total - the total number of HTTP requests received). 

In our case the metric name should be: `successful_calls`, `failed_calls`, `buffered_calls` or `state`.

Combined: `resilience4j_circuitbreaker_successful_calls`

> Use labels to differentiate the characteristics of the thing that is being measured. Do not put the label names in the metric name.

In our case labels could be: `name=<circuitBreakerName>` or  `name=<rateLimiterName>`.
Other labels could be the `application=<spring.application.name>`.

Combined: `resilience4j_circuitbreaker_successful_calls{name="Backend1", application="DEMO"}

What do you think?
If we don't put the kind/type into labels, but into the metric name, does that fulfill your requirements?

**Comment by RobWin** [comment]
I think we should agree on a naming convention and fix both `resilience4j-prometheus` and tagged `resilience4j-micrometer` so that the output looks similar.

**Comment by leonard84** [comment]
> In our case the metric name should be: successful_calls, failed_calls, buffered_calls or state.

Hm, I don't quite agree in the example you quote they use `http_requests_total` and not `http_requests_successful`, the success/status_code is just a tag and not a metric itself. 
The same would apply to the resilience4j metrics, e.g. `resilience4j_circuitbreaker_calls{name=<circuitBreakerName>, state=<successful|failed|buffered>}`

> Other labels could be the `application=<spring.application.name>`

General tags should not be added by a library, as that is normally done via a generic enrichment.

**Comment by noroute** [comment]
@leonard84 is right. Dimensions of a metric (success/failure/id/etc.) should be rendered as label. Including dimensions in the name makes discovery of dimensions impossible.

In all cases I know, application identity is added by infrastructure (e.g. Prometheus relabeling) and does not need to be included by the app unless it has special knowledge about that identity.

With some further looking into the details of resilience4js submodules I found:
`resilience4j-prometheus` already produces nice timeseries with tags:
```
# HELP resilience4j_circuitbreaker_calls Circuit Breaker Call Stats
# TYPE resilience4j_circuitbreaker_calls gauge
resilience4j_circuitbreaker_calls{name="backendB",call_result="successful",} 0.0
```
(see [Code](https://github.com/resilience4j/resilience4j/blob/1e063a4d96b259044dcb4c6fe2d64e38a8862598/resilience4j-prometheus/src/main/java/io/github/resilience4j/prometheus/CircuitBreakerExports.java#L211))

This also works nicely for the Spring Boot 1.x Starter since autoconfigure uses the [Prometheus integration directly](https://github.com/resilience4j/resilience4j/blob/master/resilience4j-spring-boot/src/main/java/io/github/resilience4j/circuitbreaker/autoconfigure/CircuitBreakerPrometheusAutoConfiguration.java). The Spring Boot 2.x Starter (correctly) only supports [autoconfigure for Micrometer](https://github.com/resilience4j/resilience4j/blob/master/resilience4j-spring-boot2/src/main/java/io/github/resilience4j/circuitbreaker/autoconfigure/CircuitBreakerMetricsAutoConfiguration.java).

If I'm not mistaken, the only thing that's broken is `resilience4j-micrometer` since it forces all dimensions into the metric name, although Micrometer could easily handle dimensions and adapt the output to the respective time series database.

In short: use the names and tags from `resilience4j-prometheus` in `resilience4j-micrometer` and be done with it. This remains a breaking change (SemVer?) but should only affect people using Spring Boot 2 or  Micrometer directly.

**Comment by RobWin** [comment]
@leonard84  @noroute 
Yes, you are right. The call results `successfull | failed | not_permitted` are comparable to HTTP status codes 200, 500 or 401/403 of http_requests and should be labels/tags.

But I actually think that `buffered` and `buffered_max` are metrics by their own. They are not call results and thus not dimensions of the metric `resilience4j_circuitbreaker_calls`.
`buffered` is the sum of successful calls and failed calls up to `buffered_max`. More like thread pool size and max pool size.
What about  `resilience4j_circuitbreaker_buffer_size` and `resilience4j_circuitbreaker_buffer_max_size`.

What do you think?

**Comment by noroute** [comment]
@RobWin Oh, thanks for the clarification. I didn't get the intricacies of buffers, yet. Is there documentation about that or do I have to look into the code?

# What CircuitBreaker metrics mean (interpretation)
Let me try to recap how this works (CircuitBreaker only):
* `successful/failed/not_permitted` are all call results (and thus dimensions of *calls*)
* `buffered_max` is the *size* of the ring buffer BitSet (and would be more appropriately called `buffer_size`) (and thus a dimension of the *circuit breaker*)
* `buffered` is the current fill ratio of the ring buffer (and thus a dimension of the *circuit breaker*, too)

`resilience4j_circuitbreaker_calls_total` is not available in any of the metrics exporters but could easily be calculated so there does not seem to be a need for this.

Is this correct?

# New design
IMHO, this leaves us with 3 metrics per circuit breaker (in Prometheus notation for clarity):
```
resilience4j_circuitbreaker_calls{name="foo", call_results="success"} 1.0
resilience4j_circuitbreaker_state{name="foo", state="closed"} 1.0
resilience4j_circuitbreaker_buffer{name="foo", value="current"} 2.0
resilience4j_circuitbreaker_buffer{name="foo", value="max"} 10.0
```

# Consequences
1. This would be a breaking change for *all* metrics backends since it splits off the buffer metrics. This would not be too bad since it syncs handling for all backends and would be a clean cut.
2. We should also adapt all other metrics (non-CircuitBreaker) to the same style.

What do you think?

**Comment by RobWin** [comment]
We document some implementation details here: http://resilience4j.github.io/resilience4j/#_circuitbreaker_11

**Comment by RobWin** [comment]
If I would be picky, I would say that `not_permitted` is also no call result.
If the CircuitBreaker is OPEN it does not permit any calls, because calls are blocked. 
That means: Not permitted calls are only measured when the CircuitBreaker is OPEN. As soon as it transitions back to CLOSED, the number of permitted calls are resetted.

**Comment by RobWin** [comment]
To answer your question regarding the consequences.
1.) I think it would be good to "fix" and align the metrics output if r4j-micrometer and r4j-prometheus. Even if it is a non-backward compatible change.
2.) Yes, we should adapt all other metrics to the same style.

**Comment by RobWin** [comment]
I still don't get why metric names should be inside of labels:

For example
```
resilience4j_circuitbreaker_buffer{name="foo", value="current"} 2.0
resilience4j_circuitbreaker_buffer{name="foo", value="max"} 10.0
```

vs 
`` 
resilience4j_circuitbreaker_buffered_calls{name="foo"} 2.0
resilience4j_circuitbreaker_max_buffered_calls{name="foo"} 10.0
```

Maybe I'm too used to InfluxDB. where you can have multiple fields and tags per measurement. And tags can be used to query measurements.

**Comment by leonard84** [comment]
> If I would be picky, I would even say that not_permitted is also no call result.
> If the CircuitBreaker is OPEN it does not permit any calls, because calls are blocked.
> That means: Not permitted calls are only measured when the CircuitBreaker is OPEN. As soon as it transitions back to CLOSED, the number of permitted calls are resetted.

What I noticed is that a `gauge` metric is used both for calls and state, which is incorrect IMHO. Calls should be a counter, since they always increase, state is a gauge since it toggle between 0 and 1.
This would also imply that not_permitted wouldn't be reset.

**Comment by RobWin** [comment]
No, number of calls don't increase. The only increase up to the size of the ring buffer. Gauge is fine for successful and failed calls.

**Comment by leonard84** [comment]
Then it would make sense, to add counter metrics for statistics as well. E.g. I want to see the total error rate for 24h. This doesn't work for gauges.

> If it helps, think of a Gauge as a "heisen-gauge" - a meter that only changes when it is observed. Every other meter type provided out-of-the-box accumulates intermediate counts toward the point where the data is sent to the metrics backend.

https://micrometer.io/docs/concepts#_gauges

**Comment by RobWin** [comment]
[CircuitBreaker metrics](https://github.com/resilience4j/resilience4j/blob/master/resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerMetrics.java) report the current values of the internal [RingBitSet](https://github.com/resilience4j/resilience4j/blob/master/resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/RingBitSet.java).  Besides that it has a LongAdder (counter) which counts the number of not permitted calls when the CircuitBreaker is OPEN. The RingBitSet  has a upper bound which is the `ringBufferSize`.
CircuitBreaker metrics do not report any other counters like total sum of all measures calls, sucessful calls or failed calls. I don't want to add other counters to the CircuitBreaker. It's should be lightweight. , since This should be the duty of a different component.

If it's needed, you can easily add event consumers to a CircuitBreaker and count events with Micrometer or Dropwizard.

**Comment by RobWin** [comment]
Or you could use a Dropwizard Metrics decorator which does that for you.  But I think there no MicroMeter alternative yet.

You could decorate your call and get counters for successful calls, failed calls and even timings.

```
Timer timer = Timer.of("backend");
Supplier<String> supplier = Timer.decorateSupplier(timer, backendService::doSomething);

// Retrieve Timer metrics
Timer.Metrics metrics = timer.getMetrics();
// Returns the number of total calls
long totalCalls = metrics.getNumberOfTotalCalls();
// Returns the number of successful calls
long successfulCalls = metrics.getNumberOfSuccessfulCalls();
// Returns the number of failed calls
long failedCalls = metrics.getNumberOfFailedCalls();
```

**Comment by RobWin** [comment]
I updated my last two comments, don't read them in EMails ;)

**Comment by leonard84** [comment]
@RobWin I never meant that CircuitBreaker should keep a counter for all calls. What you described with the EventPublisher is close to what I had in mind. The only difference is that it should use tags instead of custom names and it should also pick up added CircuitBreakers.

**Comment by RobWin** [comment]
I just wanted to explain why a CB has no other counter metrics.

**Comment by noroute** [comment]
> I still don't get why metric names should be inside of labels:
> 
> For example
> 
> ```
> resilience4j_circuitbreaker_buffer{name="foo", value="current"} 2.0
> resilience4j_circuitbreaker_buffer{name="foo", value="max"} 10.0
> ```
> vs
> 
> ```
> resilience4j_circuitbreaker_buffered_calls{name="foo"} 2.0
> resilience4j_circuitbreaker_max_buffered_calls{name="foo"} 10.0
> ```
> Maybe I'm too used to InfluxDB and Grafana. where you can have multiple fields and tags per measurement. And tags can be used to query measurements.
> 
> Aren't labels/tags better used to define dimensions like environment, applicationName, ...

I agree that keeping current and max separate is the way it's done in Prometheus as they represent two different concepts (configured max vs. current fill). I would change the name to help discovery, though:

```
resilience4j_circuitbreaker_buffered_calls{name="foo"} 2.0
resilience4j_circuitbreaker_buffered_calls_max{name="foo"} 10.0
```

I would also change the `name` tag to `cb` to clarify that represents an identifier.

```
resilience4j_circuitbreaker_buffered_calls{cb="foo"} 2.0
resilience4j_circuitbreaker_buffered_calls_max{cb="foo"} 10.0
```

I'll do a draft PR for Micrometer CircuitBreaker so we can see it in action, but I probably won't get to that until next week. That PR should then grow to encompass all Metrics and all backends.

Objections?

**Comment by RobWin** [comment]
No, I don't have any objections anymore. Only two remarks.

`resilience4j_circuitbreaker_buffered_calls_max` could be confusing. It could be understood as the maximum value ever measured. But this value is constant and never increases or increases.
I would really call it `resilience4j_circuitbreaker_max_buffered_calls`.

I would also not use abbreviations as label names. Let's call it `circuitbreaker` or `ratelimiter`.

**Comment by leonard84** [comment]
`resilience4j_circuitbreaker_buffered_calls_max` is that value based on a config? If that is the case we might name it accordingly, e.g. `tomcat_threads_config_max_threads`, to make it clear that it is not an observation but a constant value.

**Comment by RobWin** [comment]
Yes it's the max ring buffer value based on a config

**Comment by voievodin** [comment]
>I'll do a draft PR for Micrometer CircuitBreaker so we can see it in action, but I probably won't get to that until next week. That PR should then grow to encompass all Metrics and all backends.

@noroute there is a PR hanging around with the changes related to it :point_right: #355. It introduces tag based binders. Since it was created 2 weeks ago the naming wasn't aligned with this discussion. I've just aligned it to correspond to the thoughts you guys had here, but with a few differences:
```
# cb
resilience4j_circuitbreaker_state{name="Backend1"}

resilience4j_circuitbreaker_calls{name="Backend1", kind="successful"}
resilience4j_circuitbreaker_calls{name="Backend1", kind="failed"}
resilience4j_circuitbreaker_calls{name="Backend1", kind="not_permitted"}

resilience4j_circuitbreaker_buffered_calls{name="Backend1"}
resilience4j_circuitbreaker_max_buffered_calls{name="Backend1"}

#bulkhead
resilience4j_bulkhead_concurrent_calls{name="Backend1", kind="available"}
resilience4j_bulkhead_max_available_concurrent_calls{name="Backend1"}
```

I would argue that we need 3 state gauges (per state), the current approach seems nice to me, and fits well, let's say with grafana stat view. I would also argue with the necessity to have tag duplicating metric names, e.g.
```
resilience4j_circuitbreaker_calls{circuitbreaker="Backend1", kind="successful"}
# vs
resilience4j_circuitbreaker_calls{name="Backend1", kind="successful"}
```
looks odd to me, since "circuitbreaker" is already part of the metric name. The same applies to the `call_result`, the metric itself is already showing the call result we just need to categorize it, maybe there better options than `kind` though:
```
resilience4j_circuitbreaker_calls{name="Backend1", call_result="successful"}
# vs
resilience4j_circuitbreaker_calls{name="Backend1", kind/type="successful"}
```

**Comment by noroute** [comment]
@RobWin, @voievodin, @leonard84 The draft above includes all the naming discussions we had and has a sample implementation for CircuitBreaker in micrometer. If that looks fine to you (also style-wise) I can do the rest (other Meters plus direct prometheus and Dropwizard integration.

Note that in contrast to #355 this *is* breaking backward compatibility.

**Comment by RobWin** [comment]
The PR from @voievodin looks good to me. I think I will merge his PR soon.

**Comment by noroute** [comment]
@RobWin The PR from @voievodin looks well structured so far. Note that it does require further changes in the spring-boot modules to be usable (since it introduces new metrics classes) and does *not* include the agreed-upon naming for the configured max buffer size und tag names. The tag names (`kind`) would also be inconsistent with `resilience4j-prometheus` which uses `call_result`.

PR #374 does adhere to the names and structure discussed in this issue. It is yet incomplete (CircuitBreaker only) but I'm willing to move ahead if you chose this alternative.

**Comment by RobWin** [comment]
@voievodin I merged PR #355. 
PR #375 also looks good to me. I've only some minor findings.
Could you also adapt Spring Boot and Spring Boot 2 to use tagged metrics? 
I know it's not backward compatible. Maybe we can make it configurable so that the user can decide if he wants to stick with the old implementation. The tagged implementation would be the new default.

**Comment by voievodin** [comment]
Thank you.

>Could you also adapt Spring Boot and Spring Boot 2 to use tagged metrics?

sure.

**Comment by voievodin** [comment]
Here we go @RobWin https://github.com/resilience4j/resilience4j/pull/382

**Comment by noroute** [comment]
@voievodin Thanks for the new binders and actually doing the work. Much appreciated!

**Comment by voievodin** [comment]
A pleasure to collaborate with you guys.
@RobWin do you think we can have this one released ?

**Comment by RobWin** [comment]
There are some open issues which must be solved first. At least the refactoring of our Spring AOP aspects is important.

**Comment by RobWin** [comment]
While we are wroking on the last open issues.
@voievodin would you like to have a look at  https://github.com/resilience4j/resilience4j/pull/255
Maybe it's also interesting for you.

**Comment by voievodin** [comment]
Sure, will take a look this weekend

**Comment by voievodin** [comment]
Have a [draft](https://github.com/resilience4j/resilience4j/compare/master...voievodin:elapsed-cb) =) Let's continue the discussion [here](https://github.com/resilience4j/resilience4j/pull/255#issuecomment-480106967)

**Comment by RobWin** [comment]
0.14.0 was released

**Comment by leonard84** [comment]
@voievodin / @RobWin it seems I haven't kept up with this as I should have but I don't quite agree with the decision to conflate the states for `resilience4j_circuitbreaker_state` from explicit tags to simple gauge values

    0 - CLOSED
    1 - OPEN
    2 - HALF_OPEN

While it is just a question of style when you look at a single instance. It makes a huge difference when working with multiple instances. With tags I could just say `sum(resilience4j_circuitbreaker_state{state="open"})` to get all open circuitbreakers in my cluster.

**Comment by RobWin** [comment]
I think that didn't change compared to the previous implementation.
But I see your point. 

Can you use the comparison operation `==` to filter in the meantime?

Would you like to create a PR to fix it?

**Comment by leonard84** [comment]
I've created a proposal https://github.com/resilience4j/resilience4j/pull/416 what do you think? 

/cc @voievodin 

Still todo would be to update the docs and so on, which I would do once we decide to move forward with this.
